File: packages/judgment-core/src/dependency-graph.ts
/**
 * In-memory dependency graph for tracking relationships between judgment points.
 *
 * An edge from A to B means "B depends on A" (A is upstream, B is downstream).
 * When A changes, B may become stale.
 */
export class DependencyGraph {
  /** Maps a point ID to the set of IDs that depend on it (downstream). */
  private readonly downstream = new Map<string, Set<string>>();
  /** Maps a point ID to the set of IDs it depends on (upstream). */
  private readonly upstream = new Map<string, Set<string>>();

  /**
   * Add a dependency edge: downstream depends on upstream.
   */
  addDependency(upstreamId: string, downstreamId: string): void {
    let downSet = this.downstream.get(upstreamId);
    if (downSet == null) {
      downSet = new Set();
      this.downstream.set(upstreamId, downSet);
    }
    downSet.add(downstreamId);

    let upSet = this.upstream.get(downstreamId);
    if (upSet == null) {
      upSet = new Set();
      this.upstream.set(downstreamId, upSet);
    }
    upSet.add(upstreamId);
  }

  /**
   * Remove a dependency edge.
   */
  removeDependency(upstreamId: string, downstreamId: string): void {
    this.downstream.get(upstreamId)?.delete(downstreamId);
    this.upstream.get(downstreamId)?.delete(upstreamId);
  }

  /**
   * Get the direct upstream dependencies of a point.
   */
  getUpstream(pointId: string): readonly string[] {
    return Array.from(this.upstream.get(pointId) ?? []);
  }

  /**
   * Get the direct downstream dependents of a point.
   */
  getDownstream(pointId: string): readonly string[] {
    return Array.from(this.downstream.get(pointId) ?? []);
  }

  /**
   * Get all transitive downstream dependents of a point (BFS).
   */
  getTransitiveDependents(pointId: string): readonly string[] {
    const visited = new Set<string>();
    const queue: string[] = [...(this.downstream.get(pointId) ?? [])];

    while (queue.length > 0) {
      const current = queue.shift();
      if (current == null || visited.has(current)) continue;
      visited.add(current);

      for (const next of this.downstream.get(current) ?? []) {
        if (!visited.has(next)) {
          queue.push(next);
        }
      }
    }

    return Array.from(visited);
  }

  /**
   * Check whether adding a dependency from upstream to downstream would create a cycle.
   */
  wouldCreateCycle(upstreamId: string, downstreamId: string): boolean {
    if (upstreamId === downstreamId) return true;

    // Would create cycle if downstreamId can transitively reach upstreamId.
    const visited = new Set<string>();
    const queue: string[] = [...(this.downstream.get(downstreamId) ?? [])];

    while (queue.length > 0) {
      const current = queue.shift();
      if (current == null) continue;
      if (current === upstreamId) return true;
      if (visited.has(current)) continue;
      visited.add(current);

      for (const next of this.downstream.get(current) ?? []) {
        if (!visited.has(next)) {
          queue.push(next);
        }
      }
    }

    return false;
  }

  /**
   * Remove all edges associated with a point.
   */
  removePoint(pointId: string): void {
    // Remove from downstream neighbors' upstream sets.
    for (const downId of this.downstream.get(pointId) ?? []) {
      this.upstream.get(downId)?.delete(pointId);
    }
    // Remove from upstream neighbors' downstream sets.
    for (const upId of this.upstream.get(pointId) ?? []) {
      this.downstream.get(upId)?.delete(pointId);
    }
    this.downstream.delete(pointId);
    this.upstream.delete(pointId);
  }
}

File: backend/judgment_core/src/judgment_core/dependency_graph.py
"""In-memory dependency graph for tracking relationships between judgment points.

An edge from A to B means "B depends on A" (A is upstream, B is downstream).
When A changes, B may become stale.
"""

from __future__ import annotations


class DependencyGraph:
    """Tracks directed dependencies between judgment points."""

    def __init__(self) -> None:
        self._downstream: dict[str, set[str]] = {}
        self._upstream: dict[str, set[str]] = {}

    def add_dependency(self, upstream_id: str, downstream_id: str) -> None:
        """Add a dependency edge: *downstream_id* depends on *upstream_id*."""
        self._downstream.setdefault(upstream_id, set()).add(downstream_id)
        self._upstream.setdefault(downstream_id, set()).add(upstream_id)

    def remove_dependency(self, upstream_id: str, downstream_id: str) -> None:
        """Remove a dependency edge."""
        down_set = self._downstream.get(upstream_id)
        if down_set is not None:
            down_set.discard(downstream_id)
        up_set = self._upstream.get(downstream_id)
        if up_set is not None:
            up_set.discard(upstream_id)

    def get_upstream(self, point_id: str) -> list[str]:
        """Get the direct upstream dependencies of a point."""
        return list(self._upstream.get(point_id, set()))

    def get_downstream(self, point_id: str) -> list[str]:
        """Get the direct downstream dependents of a point."""
        return list(self._downstream.get(point_id, set()))

    def get_transitive_dependents(self, point_id: str) -> list[str]:
        """Get all transitive downstream dependents of a point (BFS)."""
        visited: set[str] = set()
        queue = list(self._downstream.get(point_id, set()))

        while queue:
            current = queue.pop(0)
            if current in visited:
                continue
            visited.add(current)
            for next_id in self._downstream.get(current, set()):
                if next_id not in visited:
                    queue.append(next_id)

        return list(visited)

    def would_create_cycle(self, upstream_id: str, downstream_id: str) -> bool:
        """Check whether adding a dependency would create a cycle."""
        if upstream_id == downstream_id:
            return True

        visited: set[str] = set()
        queue = list(self._downstream.get(downstream_id, set()))

        while queue:
            current = queue.pop(0)
            if current == upstream_id:
                return True
            if current in visited:
                continue
            visited.add(current)
            for next_id in self._downstream.get(current, set()):
                if next_id not in visited:
                    queue.append(next_id)

        return False

    def remove_point(self, point_id: str) -> None:
        """Remove all edges associated with a point."""
        for down_id in list(self._downstream.get(point_id, set())):
            up_set = self._upstream.get(down_id)
            if up_set is not None:
                up_set.discard(point_id)

        for up_id in list(self._upstream.get(point_id, set())):
            down_set = self._downstream.get(up_id)
            if down_set is not None:
                down_set.discard(point_id)

        self._downstream.pop(point_id, None)
        self._upstream.pop(point_id, None)
